### ISSUE TITLE
twisted.spread.banana.decode is stateful

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2319]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2319|
|Type|defect|
|Created|2006-12-21 01:08:09Z|

```
>>> from twisted.spread.banana import encode, decode
>>> decode(encode(''))
''
>>> decode(encode(2 ** 512))
Traceback (most recent call last):
  File "<stdin>", line 1, in ?
  File "/home/exarkun/Projects/Twisted/trunk/twisted/spread/banana.py", line 339, in decode
    _i.dataReceived(st)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/spread/banana.py", line 138, in dataReceived
    raise BananaError("Security precaution: longer than 64 bytes worth of prefix")
twisted.spread.banana.BananaError: Security precaution: longer than 64 bytes worth of prefix
>>> decode(encode(''))
Traceback (most recent call last):
  File "<stdin>", line 1, in ?
  File "/home/exarkun/Projects/Twisted/trunk/twisted/spread/banana.py", line 339, in decode
    _i.dataReceived(st)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/spread/banana.py", line 138, in dataReceived
    raise BananaError("Security precaution: longer than 64 bytes worth of prefix")
twisted.spread.banana.BananaError: Security precaution: longer than 64 bytes worth of prefix
>>> 
```

I'm pretty sure the 2nd call to ```decode(encode(''))``` should work, not raise a BananaError.

<details><summary>Searchable metadata</summary>

```
trac-id__2319 2319
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1166663289000000 1166663289000000
changetime__1166986452000000 1166986452000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

Fixed in stateful-decode-2319

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_2"></a>|@jerub commented|
|-|-|

Thankyou for the test and the fix.

I think using a method of .append in 'decode' is just nasty, but wow, I guess there isn't another way that doesn't involve a bunch more code for no benefit.

Please merge this.

Further, and independant of my review of the code in question, I discovered this line of code, from twisted.spread.banana:
```
                    log.msg('freaky losing')
```
I would prefer that crappy log messages like this didn't exist. Perhaps a ticket should be filed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

(In [19066]) Merge stateful-decode-2319

Author: exarkun
Reviewer: jerub
Fixes [#2319](https://github.com/twisted/twisted/issues/2319)

Add a test for the behavior of twisted.spread.banana.decode after being called
with a string which it cannot decode.  Fix its behavior so that this occurrance
does not permanently corrupt its state preventing all future calls from succeeding.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
